crates/ide-assists/src/handlers/remove_parentheses.rs

```diff
@@ -0,0 +1,198 @@
+use syntax::{ast, AstNode, SyntaxKind, TextRange};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: remove_parentheses
+//
+// Removes useless parentheses.
+//
+// ```
+// fn main() {
+//     _ = $0(2) + 2;
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     _ = 2 + 2;
+// }
+// ```
+pub(crate) fn remove_parentheses(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let parens = ctx.find_node_at_offset::<ast::ParenExpr>()?;
+    let l_paren = parens.l_paren_token()?;
+    let r_paren = parens.r_paren_token()?;
+
+    let cursor_in_range = l_paren.text_range().contains_range(ctx.selection_trimmed())
+        || r_paren.text_range().contains_range(ctx.selection_trimmed());
+    if !cursor_in_range {
+        return None;
+    }
+
+    let expr = parens.expr()?;
+    let parent = ast::Expr::cast(parens.syntax().parent()?);
+    let is_ok_to_remove =
+        parent.map_or(true, |p| ExprPrecedence::of(&expr) >= ExprPrecedence::of(&p));
+    if !is_ok_to_remove {
+        return None;
+    }
+
+    let delete_from_l = l_paren.text_range().start();
+    let delete_to_l = match l_paren.next_token() {
+        Some(it) if it.kind() == SyntaxKind::WHITESPACE => it.text_range().end(),
+        _ => l_paren.text_range().end(),
+    };
+
+    let delete_from_r = match r_paren.prev_token() {
+        Some(it) if it.kind() == SyntaxKind::WHITESPACE => it.text_range().start(),
+        _ => r_paren.text_range().start(),
+    };
+    let delete_to_r = r_paren.text_range().end();
+
+    let target = parens.syntax().text_range();
+    acc.add(
+        AssistId("remove_parentheses", AssistKind::Refactor),
+        "Remove parentheses",
+        target,
+        |builder| {
+            builder.delete(TextRange::new(delete_from_l, delete_to_l));
+            builder.delete(TextRange::new(delete_from_r, delete_to_r));
+        },
+    )
+}
+
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
+pub enum ExprPrecedence {
```

Comment:
There should be a better place for all of this, where should I move precedence handling?